[Tee Chu Jie] iP

tamagochuuu opened the pull request.

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

tamagochuuu committed:
Finished level-0

tamagochuuu committed:
edited level 0

tamagochuuu committed:
finished level 1

tamagochuuu committed:
finished level 2

tamagochuuu committed:
finished level 3

tamagochuuu committed:
finished Level 4

tamagochuuu committed:
Added A-TextUiTesting

tamagochuuu committed:
Level 5

tamagochuuu committed:
Level 6 and javadocs

tamagochuuu committed:
add save function

Javiery3889 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,378 @@
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.Scanner;
+import java.util.List;
```
I like how you have explicitly listed out the imported classes, adhering to our coding standards 👍

Javiery3889 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,378 @@
+    /**
+     * Initiating the user's todo list.
+     */
+    static ArrayList<Task> list = new ArrayList<>();
```
Perhaps a more intuitive variable name could be used here like `todoList`?

Javiery3889 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,378 @@
+
+                case "deadline":
+                    deadline();
+                    break;
+
+                case "todo":
+                    todo();
+                    break;
+
+                case "event":
+                    event();
+                    break;
+
+                case "list":
+                    list();
+                    break;
+
+                case "mark":
+                    mark();
+                    break;
+
+                case "unmark":
+                    unmark();
+                    break;
+
+                case "delete":
+                    delete();
+                    break;
+
+                case "bye":
+                    try {
+                        save("list.txt");
+                    } catch(IOException ex) {
+                        System.out.println("Something went wrong: " + ex.getMessage());
+                    }
+                    bye();
+
+                default:
+                    break;
+            }
```
Following our coding standards, `case` clauses **should not** have any indentation in switch statements

Javiery3889 commented on src/main/java/DukeException.java:
```diff
@@ -0,0 +1,28 @@
+     * @param command The command to be checked.
+     * @throws DukeException The exception that will be thrown if command is invalid.
+     */
+    public static void validate(String command) throws DukeException {
+        if (!Duke.commands.contains(command)) {
+            throw new DukeException("Invalid Command Given");
+        }
+    }
```
Any reason why you did user input validation in `DukeException.java` instead of `Duke.java`? Based on the textbook, exceptions should only be used for "unusual" conditions instead of normal control flow. Consider separating the validation method and throwing this exception after validation instead.

Javiery3889 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,378 @@
+    /**
+     * Says goodbye to Mitsuki and exits the Chat bot.
+     */
+    public static void bye() {
```
Following our coding standards, method names must be verbs. Perhaps `showBye()` would be more intuitive? This issue applies for some of the method names above in this file as well.

Javiery3889 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,378 @@
+        }
+
+        // Adding the commands to the list of commands Mitsuki can execute.
+        commands.add("help");
+        commands.add("list");
+        commands.add("todo");
+        commands.add("deadline");
+        commands.add("event");
+        commands.add("mark");
+        commands.add("unmark");
+        commands.add("delete");
+        commands.add("bye");
+
+
+
+        // Initiating the 'command' variable.
+        String command = "nil";
+
+        // The loop that takes in user input and determine what to do.
+        while (!command.equals(null)) {
+            // Taking in next user input.
+            command = scan.next();
+
+            // Making sure user is giving a valid command.
+            try {
+                DukeException.validate(command);
+            } catch(DukeException ex) {
+                System.out.println("Sorry, I am unable to do that at the current moment.");
+                System.out.println("Please type 'help' for a list of commands you can give me! :)");
+            }
```
Perhaps using the `default` clause in the switch statement below to handle invalid commands would be intuitive here?

Javiery3889 reviewed (commented):
LGTM except for a few minor nits regarding coding standard violations. Consider separating the logic for parsing user input in `Duke.java` into another class as the `Duke` class is getting quite long. Press on! 👍

agreatdayy commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,378 @@
+            AlreadyDoneException.validate(theTask);
+        } catch(AlreadyDoneException ex) {
+            System.out.println("You have already completed this task! :D");
+            System.out.println(theTask);
```
Perhaps you can combine these lines together?

agreatdayy commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,378 @@
+            WasNotDoneException.validate(aTask);
+        } catch(WasNotDoneException ex) {
+            System.out.println("You had not completed this task! :O");
+            System.out.println(aTask);
```
Maybe can combine these 2 lines together too?

agreatdayy commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,378 @@
+                String description = fullString.substring(7, descriptionEnd - 1);
+                String deadline = fullString.substring(descriptionEnd + 1, deadlineEnd);
+
+                Task nextTask = new Deadline(description, deadline);
```
Could this be extracted into the Deadline constructor?

agreatdayy commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,378 @@
+                String startTime = fullString.substring(descriptionEnd + 1, startTimeEnd - 1);
+                String endTime = fullString.substring(startTimeEnd - 1, endTimeEnd);
+
+                Task nextTask = new Event(description, startTime, endTime);
```
Could this be extracted into the Event constructor too?

agreatdayy commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,32 @@
+     * Converts the Deadline object into a human-readable String to be displayed to the user.
+     * @return String object that represents the Deadline object.
+     */
+    @Override
```
Following our coding standards, `@Override` methods do not need javadocs. But good job having concise javadocs that are easy to understand!

agreatdayy commented on src/main/java/Event.java:
```diff
@@ -0,0 +1,40 @@
+     * Converts the Event object into a human-readable String to be displayed to the user.
+     * @return String object that represents the Event object.
+     */
+    @Override
```
I've noticed that you have javadocs for other `@Override` methods in other classes as well. Please refer to the comment I made in the `Deadline` class.

agreatdayy commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,378 @@
+     */
+    public static void help() {
+        System.out.println("Here is a list of commands you can give me:");
+        for (int i = 0; i < commands.size(); i++) {
+            int j = i + 1;
+            System.out.println(j + ". " + commands.get(i));
+        }
```
Would adding more details for each command make it easier for users to use the command words properly?

agreatdayy commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,378 @@
+        String fullString = scan.nextLine();
+
+        String[] tokens = fullString.split("/");
+        try {
+            EmptyTaskException.timedValidate(tokens);
+            MissingDeadlineException.validate(tokens);
+        } catch(EmptyTaskException ex) {
+            System.out.println("Please give a description for your deadline item. Try again!");
+            return;
+        } catch(MissingDeadlineException ex) {
+            System.out.println("Please give a deadline for your deadline item. Try again! XD");
+            System.out.println("E.g. Type 'deadline return book /by Sunday 2pm' to add the task 'return book' " +
+                    "\nwith a deadline of 'Sunday 2pm' to your list.");
+            return;
```
I really like how you put in extra work to catch the different errors. Good job! :)

agreatdayy commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,378 @@
+     */
+    public static void list() {
+        System.out.println("Here are the items in your list:");
+        int i = 0;
+        int j = 1;
+        while (i < list.size()) {
+            System.out.println(j + ". " + list.get(i).toString());
+            i++;
+            j++;
+        }
```
Could this be written better using a for loop instead of using a while loop and 2 variables?

agreatdayy reviewed (commented):
Overall, I found your code easy to follow but it could be even better with more OOP! You got this! 👍

tamagochuuu committed:
finish level 8

tamagochuuu committed:
Merge pull request #2 from tamagochuuu/branch-level-8

finish level 8

tamagochuuu committed:
do MoreOOP

tamagochuuu committed:
Merge pull request #3 from tamagochuuu/A-MoreOOP

do MoreOOP

tamagochuuu committed:
package the goods

tamagochuuu committed:
Merge pull request #4 from tamagochuuu/A-Packages

package the goods

tamagochuuu committed:
Merge pull request #5 from tamagochuuu/add-gradle-support

Add gradle support

tamagochuuu committed:
add gradle support

tamagochuuu committed:
add junit tests

tamagochuuu committed:
Merge pull request #6 from tamagochuuu/A-JUnit

add junit tests

tamagochuuu committed:
Make jar file

tamagochuuu committed:
Merge pull request #7 from tamagochuuu/A-Jar

Make jar file

tamagochuuu committed:
add javadocs

tamagochuuu committed:
Merge pull request #8 from tamagochuuu/A-JavaDoc

add javadocs

tamagochuuu committed:
comply to coding standards

tamagochuuu committed:
Merge pull request #9 from tamagochuuu/A-CodingStandard

comply to coding standards

tamagochuuu committed:
implement find

tamagochuuu committed:
Merge pull request #10 from tamagochuuu/Level-9

implement find

tamagochuuu committed:
add level-10 features, not debugged

tamagochuuu committed:
Merge pull request #11 from tamagochuuu/branch-level-10

add level-10 features, not debugged

tamagochuuu committed:
debug level 10

tamagochuuu committed:
Merge pull request #12 from tamagochuuu/debug-level-10

debug level 10

tamagochuuu committed:
create new jar file

tamagochuuu committed:
create new jar file

tamagochuuu committed:
Merge pull request #13 from tamagochuuu/new-jar-file

Create new .jar file

The current executable .jar file does not include the JavaFX library, and the chatbot
will not run properly on computers that do not have JavaFX installed.

Create a new executable .jar file using gradle to allow Mitsuki to be run
on computers that do not have JavaFX installed.

tamagochuuu committed:
Create new .jar file

The current executable .jar file does not include the JavaFX library, and the chatbot
will not run properly on computers that do not have JavaFX installed.

Create a new executable .jar file using gradle to allow Mitsuki to be run
on computers that do not have JavaFX installed.

tamagochuuu committed:
Merge pull request #14 from tamagochuuu/newjarfile

Create new .jar file

tamagochuuu committed:
Add assertions to document important assumptions

Current code does not use any method to ensure that important
assumptions hold at various points.

Add Java assertions to make sure these assumptions are true
to prevent unnecessary bugs.

tamagochuuu committed:
Merge pull request #15 from tamagochuuu/assertions

Add assertions to document important assumptions

tamagochuuu committed:
Improve code quality

Code is messy and contains undocumented public methods, as well as
redundant methods.

Rearrange parts of the code and document all public methods with
JavaDocs, and remove redundant methods.

Make the code cleaner and easier to understand.

tamagochuuu committed:
Merge pull request #16 from tamagochuuu/code-quality

Improve code quality

tamagochuuu committed:
Add the extension B-FixedDurationTask

tamagochuuu committed:
Merge pull request #17 from tamagochuuu/BCD-Extension

Add the extension B-FixedDurationTask

tamagochuuu committed:
add userguide

tamagochuuu committed:
Merge pull request #18 from tamagochuuu/A-UserGuide

add userguide

tamagochuuu committed:
create new jar file

tamagochuuu committed:
Merge pull request #19 from tamagochuuu/A-Release

create new jar file
